[Cpp Runtime] MSVC: std::min in Guid.cpp needs `NOMINMAX` macro for compilation

I am using the cpp runtime and trying to upgrade the [antlr port in vcpkg](https://github.com/microsoft/vcpkg/tree/master/ports/antlr4) from 4.9.1 to 4.9.3 (see the PR here, https://github.com/microsoft/vcpkg/pull/22438)

My dev and production environment are macOS and Linux, but the vcpkg CI checks run against with Windows environment, and when authoring the vcpkg port for 4.9.3, I found one build failure on Windows. MSVC reports errors like below:
```
[133/300] C:\PROGRA~2\MICROS~3\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\cl.exe   /TP -Dantlr4_shared_EXPORTS -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\atn -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\dfa -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\misc -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\tree -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\tree\pattern -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\tree\xpath /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /GR /EHsc /MP    /W4 /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  /Od /Zi /MP   /W4 /wd4251 -DANTLR4CPP_EXPORTS -MP /wd4251 /showIncludes /Foruntime\CMakeFiles\antlr4_shared.dir\src\support\Guid.cpp.obj /Fdruntime\CMakeFiles\antlr4_shared.dir\ /FS -c D:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support\Guid.cpp
FAILED: runtime/CMakeFiles/antlr4_shared.dir/src/support/Guid.cpp.obj 
C:\PROGRA~2\MICROS~3\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\cl.exe   /TP -Dantlr4_shared_EXPORTS -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\atn -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\dfa -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\misc -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\tree -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\tree\pattern -ID:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\tree\xpath /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /GR /EHsc /MP    /W4 /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  /Od /Zi /MP   /W4 /wd4251 -DANTLR4CPP_EXPORTS -MP /wd4251 /showIncludes /Foruntime\CMakeFiles\antlr4_shared.dir\src\support\Guid.cpp.obj /Fdruntime\CMakeFiles\antlr4_shared.dir\ /FS -c D:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support\Guid.cpp
cl : Command line warning D9025 : overriding '/W3' with '/W4'
cl : Command line warning D9025 : overriding '/Z7' with '/Zi'
D:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support\Guid.cpp(75): error C2589: '(': illegal token on right side of '::'
D:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support\Guid.cpp(75): error C2062: type 'unknown-type' unexpected
D:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support\Guid.cpp(75): error C2059: syntax error: ')'
D:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support\Guid.cpp(211): warning C4838: conversion from 'int' to 'const unsigned char' requires a narrowing conversion
D:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support\Guid.cpp(212): warning C4838: conversion from 'int' to 'const unsigned char' requires a narrowing conversion
D:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support\Guid.cpp(214): warning C4838: conversion from 'int' to 'const unsigned char' requires a narrowing conversion
D:\buildtrees\antlr4\src\4.9.3-ad469b0542.clean\runtime\src\support\Guid.cpp(215): warning C4838: conversion from 'int' to 'const unsigned char' requires a narrowing conversion
```

It looks like the usage of `std::min` in Guid.cpp makes MSVC unhappy, and MSVC considers "min" as a macro and do some unexpected pre-processing and caused this issue. Currently, I follow the approach [here](https://stackoverflow.com/a/55302808/84540) to define a `NOMINMAX` macro like this to make the vcpkg CI work for Windows MSVC:
```
 #ifdef GUID_WINDOWS
+#define NOMINMAX
 #include <objbase.h>
 #endif
```

But I see other usages of `std::min` in cpp runtime, which doesn't seem to trigger such a compilation error. I am not entirely sure if the current fix in the vcpkg is best (although it seems to work), could any one please suggest what the best approach for fixing this issue? Thanks.

--- a/runtime/Cpp/runtime/src/antlr4-common.h
+++ b/runtime/Cpp/runtime/src/antlr4-common.h
@@ -49,7 +49,7 @@
     #endif
   #endif
 
-  #define GUID_WINDOWS
+  #define ANTLR4CPP_GUID_WINDOWS 1
 
   #ifdef _WIN64
     typedef __int64 ssize_t;
@@ -68,14 +68,14 @@
   #endif
 
 #elif defined(__APPLE__)
-  #define GUID_CFUUID
+  #define ANTLR4CPP_GUID_CFUUID 1
   #if __GNUC__ >= 4
     #define ANTLR4CPP_PUBLIC __attribute__ ((visibility ("default")))
   #else
     #define ANTLR4CPP_PUBLIC
   #endif
 #else
-  #define GUID_LIBUUID
+  #define ANTLR4CPP_GUID_LIBUUID 1
   #if __GNUC__ >= 6
     #define ANTLR4CPP_PUBLIC __attribute__ ((visibility ("default")))
   #else
@@ -84,9 +84,9 @@
 #endif
 
 #ifdef __has_builtin
-#define ANTLR4_HAVE_BUILTIN(x) __has_builtin(x)
+#define ANTLR4CPP_HAVE_BUILTIN(x) __has_builtin(x)
 #else
-#define ANTLR4_HAVE_BUILTIN(x) 0
+#define ANTLR4CPP_HAVE_BUILTIN(x) 0
 #endif
 
 #include "support/Guid.h"

--- a/runtime/Cpp/runtime/src/misc/MurmurHash.cpp
+++ b/runtime/Cpp/runtime/src/misc/MurmurHash.cpp
@@ -25,7 +25,7 @@ using namespace antlr4::misc;
 #define ROTL32(x,y)	_rotl(x,y)
 #define ROTL64(x,y)	_rotl64(x,y)
 
-#elif ANTLR4_HAVE_BUILTIN(__builtin_rotateleft32) && ANTLR4_HAVE_BUILTIN(__builtin_rotateleft64)
+#elif ANTLR4CPP_HAVE_BUILTIN(__builtin_rotateleft32) && ANTLR4CPP_HAVE_BUILTIN(__builtin_rotateleft64)
 
 #define ROTL32(x, y) __builtin_rotateleft32(x, y)
 #define ROTL64(x, y) __builtin_rotateleft64(x, y)

--- a/runtime/Cpp/runtime/src/support/Guid.cpp
+++ b/runtime/Cpp/runtime/src/support/Guid.cpp
@@ -24,21 +24,27 @@
 
 #include <algorithm>
 
-#include "Guid.h"
+#include "support/Guid.h"
 
-#ifdef GUID_LIBUUID
+#ifdef ANTLR4CPP_GUID_LIBUUID
 #include <uuid/uuid.h>
 #endif
 
-#ifdef GUID_CFUUID
+#ifdef ANTLR4CPP_GUID_CFUUID
 #include <CoreFoundation/CFUUID.h>
 #endif
 
-#ifdef GUID_WINDOWS
+#ifdef ANTLR4CPP_GUID_WINDOWS
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN 1
+#endif
+#ifndef NOMINMAX
+#define NOMINMAX 1
+#endif
 #include <objbase.h>
 #endif
 
-#ifdef GUID_ANDROID
+#ifdef ANTLR4CPP_GUID_ANDROID
 #include <jni.h>
 #endif
 
@@ -153,7 +159,7 @@ std::string Guid::toString() const {
 
 // This is the linux friendly implementation, but it could work on other
 // systems that have libuuid available
-#ifdef GUID_LIBUUID
+#ifdef ANTLR4CPP_GUID_LIBUUID
 Guid GuidGenerator::newGuid()
 {
   uuid_t id;
@@ -163,7 +169,7 @@ Guid GuidGenerator::newGuid()
 #endif
 
 // this is the mac and ios version
-#ifdef GUID_CFUUID
+#ifdef ANTLR4CPP_GUID_CFUUID
 Guid GuidGenerator::newGuid()
 {
   auto newId = CFUUIDCreate(NULL);
@@ -194,7 +200,7 @@ Guid GuidGenerator::newGuid()
 #endif
 
 // obviously this is the windows version
-#ifdef GUID_WINDOWS
+#ifdef ANTLR4CPP_GUID_WINDOWS
 Guid GuidGenerator::newGuid()
 {
   GUID newId;
@@ -228,7 +234,7 @@ Guid GuidGenerator::newGuid()
 #endif
 
 // android version that uses a call to a java api
-#ifdef GUID_ANDROID
+#ifdef ANTLR4CPP_GUID_ANDROID
 GuidGenerator::GuidGenerator(JNIEnv *env)
 {
   _env = env;

--- a/runtime/Cpp/runtime/src/support/Guid.h
+++ b/runtime/Cpp/runtime/src/support/Guid.h
@@ -32,7 +32,9 @@
 #include <string>
 #include <vector>
 
-#ifdef GUID_ANDROID
+#include "antlr4-common.h"
+
+#ifdef ANTLR4CPP_GUID_ANDROID
 #include <jni.h>
 #endif
 
@@ -133,7 +135,7 @@ private:
 // each platform, but the use of newGuid is uniform.
 class GuidGenerator final {
 public:
-#ifdef GUID_ANDROID
+#ifdef ANTLR4CPP_GUID_ANDROID
   GuidGenerator(JNIEnv *env);
 #else
   GuidGenerator() { }
@@ -141,7 +143,7 @@ public:
 
   Guid newGuid();
 
-#ifdef GUID_ANDROID
+#ifdef ANTLR4CPP_GUID_ANDROID
 private:
   JNIEnv *_env;
   jclass _uuidClass;

